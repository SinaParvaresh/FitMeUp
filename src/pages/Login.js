import React, { Fragment, useState, useContext } from "react";
import { Link, useNavigate } from "react-router-dom";
import Button from "../components/UI/Button";
import classes from "./Login.module.css";
import Card from "../components/UI/Card";
import AuthContext from "../components/store/auth-context";
import { HeaderMegaMenu } from "../components/Layout/HeaderMegaMenu";

const Login = (props) => {
  const [enteredEmail, setEmail] = useState("");
  const [enteredPassword, setPassword] = useState("");
  const [isLoading, setIsLoading] = useState(false);
  const [validPassword, setValidPassword] = useState(true);
  const [emailExists, setEmailExists] = useState(true);
  const [tooManyRequests, setTooManyRequests] = useState(false);
  const navigate = useNavigate();

  const authCtx = useContext(AuthContext);

  const url =
    "https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=AIzaSyAb5ucDahLmDupsP3s5M2aSP3Hfczz-_OE";

  const submitHandler = async (event) => {
    event.preventDefault();

    if (enteredEmail.trim().length === 0 || enteredPassword.trim().length < 6) {
      return;
    }

    // Reset states
    setIsLoading(true);
    setValidPassword(true);
    setEmailExists(true);
    setTooManyRequests(false);
    console.log(enteredEmail);
    console.log(enteredPassword);

    try {
      const request = await fetch(url, {
        method: "POST",
        body: JSON.stringify({
          email: enteredEmail,
          password: enteredPassword,
          returnSecureToken: true,
        }),
      });

      const response = await request.json();
      setIsLoading(false);
      if (!response.error) {
        authCtx.login(response.idToken, Date.now() + response.expiresIn * 1000);
        navigate("/home-page");
      } else {
        let errorMessage = response.error.message;

        if (errorMessage === "EMAIL_NOT_FOUND") {
          setEmailExists(false);
        } else if (errorMessage === "INVALID_PASSWORD") {
          setValidPassword(false);
        } else if (errorMessage.substring(0, errorMessage.indexOf(" ")) === "TOO_MANY_ATTEMPTS_TRY_LATER") {
          setTooManyRequests(true);
        }
        return;
      }
    } catch (err) {
      console.log(err);
    }
  };

  const emailHandler = (event) => {
    setEmail(event.target.value.trim());
  };

  const passwordHandler = (event) => {
    setPassword(event.target.value);
  };

  return (
    <Fragment>
      <HeaderMegaMenu />
      <Card onSubmit={submitHandler}>
        <h2>Login</h2>
        {!emailExists && <div className={classes.muiAlert}>The Email is not a valid Email address</div>}
        {!validPassword && <div className={classes.muiAlert}>Incorrect password</div>}
        {tooManyRequests && <div className={classes.muiAlert}>Too many requests. Please try again later.</div>}
        <input
          className={classes.input}
          value={enteredEmail}
          onChange={emailHandler}
          type="email"
          placeholder="E-mail.com"
        />
        <input
          className={classes.input}
          value={enteredPassword}
          onChange={passwordHandler}
          type="password"
          placeholder="Password"
        />
        <Button type="submit" disabled={isLoading ? true : false}>
          Log In
        </Button>
        <div className={classes.signingActivity}>
          <Link className={classes.forgotPass} to="/forgot-password">
            Forgot Password?
          </Link>
          <Link className={classes.signUp} to="/register">
            Sign up
          </Link>
        </div>
      </Card>
    </Fragment>
  );
};

export default Login;
