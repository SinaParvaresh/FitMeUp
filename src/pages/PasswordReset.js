import React, { Fragment, useState, useRef } from "react";
import FormCard from "../components/UI/FormCard";
import classes from "./PasswordReset.module.css";
import { Button, Center, Group, Stack, Text, Title } from "@mantine/core";
import { HeaderMegaMenu } from "../components/Layout/HeaderMegaMenu";
import { FloatingLabelInput } from "../components/UI/FloatingLabelInput";
import { IconArrowLeft } from "@tabler/icons";
import { Link } from "react-router-dom";

const Register = () => {
  const forgotPassRef = useRef();
  const [enteredEmail, setEmail] = useState("");
  const [emailFound, setEmailFound] = useState(true);
  const [emailSent, setEmailSent] = useState(false);

  const url =
    "https://identitytoolkit.googleapis.com/v1/accounts:sendOobCode?key=AIzaSyAb5ucDahLmDupsP3s5M2aSP3Hfczz-_OE";

  const resetPasswordHandler = async (event) => {
    event.preventDefault();

    try {
      const request = await fetch(url, {
        method: "POST",
        body: JSON.stringify({
          requestType: "PASSWORD_RESET",
          email: enteredEmail,
        }),
        headers: {
          "Content-Type": "applicaton/json",
        },
      });
      const response = await request.json();

      if (!response.error) {
        setEmailFound(true);
        setEmailSent(true);

        return;
      } else {
        let errorMessage = response.error.message;

        if (errorMessage === "EMAIL_NOT_FOUND") {
          setEmailFound(false);
        }
      }
    } catch (err) {
      console.log(err);
    }
  };

  const emailHandler = (event) => {
    setEmail(event.target.value);
  };

  return (
    <Fragment>
      <HeaderMegaMenu></HeaderMegaMenu>
      <FormCard onSubmit={resetPasswordHandler}>
        <Stack align="center">
          {!emailSent && <Title order={3}>Forgot Password?</Title> && (
            <Text c="dimmed">Please enter the email address associated with your account.</Text>
          )}
        </Stack>

        {!emailFound && <span className={classes.failed}>Email not found</span>}
        {!emailSent && (
          <FloatingLabelInput
            type="email"
            placeholder="Email"
            label="Email"
            className={classes.input}
            value={enteredEmail}
            onChangeHandler={emailHandler}
            innerRef={forgotPassRef}
          />
        )}
        {emailSent && <span className={classes.success}>Password reset request sent! </span>}

        <Group position="apart">
          <Link to="/login" className={classes.link}>
            <Center inline>
              <IconArrowLeft stroke={1.5} size={20} />
              <Text ml={5}>Back to login page</Text>
            </Center>
          </Link>

          <Button type="submit" disabled={!enteredEmail || emailSent}>
            Submit
          </Button>
        </Group>
      </FormCard>
    </Fragment>
  );
};

export default Register;
