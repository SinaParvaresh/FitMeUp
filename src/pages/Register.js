import React, { Fragment, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import Card from "../components/UI/Card";
import classes from "./Register.module.css";
import Button from "../components/UI/Button";
import Header from "../components/Layout/Header";
import ErrorOutput from "../components/UI/ErrorOutput";

const Register = () => {
  const [enteredFirstName, setFirstName] = useState("");
  const [enteredLastName, setLastName] = useState("");
  const [enteredEmail, setEmail] = useState("");
  const [enteredPassword, setPassword] = useState("");
  const [isLoading, setIsLoading] = useState(false);
  // const [credentialError, setCredentialError] = useState(false);
  // const [emailExists, setEmailExists] = useState(false);
  // const [missingEmail, setMissingEmail] = useState(false);
  // const [weakPassword, setWeakPassword] = useState(false);
  // const [missingPassword, setMissingPassword] = useState(false);

  const [errorValidations, setErrorValidations] = useState({
    credentialError: false,
    emailExists: false,
    missingEmail: false,
    weakPassword: false,
    missingPassword: false,
  });

  const navigate = useNavigate();

  const submitHandler = (event) => {
    event.preventDefault();

    if (enteredFirstName.trim().length === 0 || enteredLastName.trim().length === 0) {
      console.log("ERROR IN VALIDATIONS");
      return;
    }

    setIsLoading(true);

    fetch("https://identitytoolkit.googleapis.com/v1/accounts:signUp?key=AIzaSyAb5ucDahLmDupsP3s5M2aSP3Hfczz-_OE", {
      method: "POST",
      body: JSON.stringify({
        email: enteredEmail,
        password: enteredPassword,
        returnSecureToken: true,
      }),
      headers: {
        "Content-type": "application/json",
      },
    }).then((res) => {
      setIsLoading(false);
      // setEmailExists(false);
      // setWeakPassword(false);
      // setMissingEmail(false);
      // setMissingPassword(false);

      if (res.ok) {
        // setCredentialError(false);
        setErrorValidations({
          credentialError: false,
        });

        navigate("/login");
      } else {
        return res.json().then((data) => {
          let errorMessage = data.error.message;
          console.log(data.error.message);

          // setCredentialError(true);

          setErrorValidations({
            credentialError: true,
          });

          if (errorMessage === "EMAIL_EXISTS") {
            // setEmailExists(true);

            setErrorValidations((prevState) => {
              return { ...prevState, emailExists: true };
            });
          } else if (errorMessage === "MISSING_EMAIL" || errorMessage === "INVALID_EMAIL") {
            // setMissingEmail(true);
            setErrorValidations((prevState) => {
              return { ...prevState, missingEmail: true };
            });
          } else if (errorMessage.substring(0, errorMessage.indexOf(" ")) === "WEAK_PASSWORD") {
            // setWeakPassword(true);
            setErrorValidations((prevState) => {
              return { ...prevState, weakPassword: true };
            });
          } else if (errorMessage === "MISSING_PASSWORD") {
            // setMissingPassword(true);
            setErrorValidations((prevState) => {
              return { ...prevState, missingPassword: true };
            });
          }
        });
      }
    });
  };

  const firstNameHandler = (event) => {
    setFirstName(event.target.value.trim());
  };

  const lastNameHandler = (event) => {
    setLastName(event.target.value.trim());
    console.log("Entered Last name:", enteredLastName);
  };

  const emailHandler = (event) => {
    setEmail(event.target.value.trim());
    console.log("Entered Email:", enteredEmail);
  };

  const passwordHandler = (event) => {
    setPassword(event.target.value);
    console.log("Entered Password:", enteredPassword);
  };

  return (
    <Fragment>
      <Header></Header>
      <Card onSubmit={submitHandler}>
        <ErrorOutput validationCheck={errorValidations} />
        {enteredFirstName.trim().length === 0 && <span>invalid name</span>}
        <input
          className={`${classes.input} ${enteredFirstName.trim().length === 0 && classes.invalid}`}
          type="firstName"
          placeholder="First Name"
          onChange={firstNameHandler}
        />
        <input className={classes.input} type="lastName" placeholder="Last Name" onChange={lastNameHandler} />
        <input
          className={`${classes.input} ${
            (errorValidations.emailExists || errorValidations.missingEmail) && classes.invalid
          }`}
          type="email"
          placeholder="E-Mail"
          onChange={emailHandler}
        />
        <input
          className={`${classes.input} ${
            (errorValidations.missingPassword || errorValidations.weakPassword) && classes.invalid
          }`}
          type="password"
          placeholder="Password"
          onChange={passwordHandler}
        />
        <Button type="submit" disabled={isLoading ? true : false}>
          Sign Up
        </Button>
        <div className={classes.signIn}>
          Have an account?
          <Link className={classes.signInLink} to="/login">
            Sign in
          </Link>
        </div>
      </Card>
    </Fragment>
  );
};

export default Register;
