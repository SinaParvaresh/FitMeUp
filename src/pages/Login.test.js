import userEvent from "@testing-library/user-event";
import { create } from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { render, screen, waitFor } from "@testing-library/react";
import React from "react";
import Login from "./Login";
import PasswordReset from "./PasswordReset";
import Register from "./Register";

describe("Login page", () => {
  test("Login Page renders correctly", () => {
    const component = create(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  test("Component routes Forgot Password link to PasswordReset component", async () => {
    const user = userEvent.setup();

    render(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );

    const component = create(
      <MemoryRouter>
        <PasswordReset />
      </MemoryRouter>
    ).toJSON();

    user.click(
      screen.getByRole("link", {
        name: /forgot password/i,
      })
    );

    await waitFor(() => {
      expect(component).toMatchSnapshot();
    });
  });

  test("Component routes Sign up link to Register component", async () => {
    const user = userEvent.setup();

    render(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );

    const component = create(
      <MemoryRouter>
        <Register />
      </MemoryRouter>
    ).toJSON();

    user.click(
      screen.getByRole("link", {
        name: /sign up/i,
      })
    );

    await waitFor(() => {
      expect(component).toMatchSnapshot();
    });
  });
});
