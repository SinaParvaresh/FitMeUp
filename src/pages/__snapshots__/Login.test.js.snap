// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page Component routes Forgot Password link to PasswordReset component 1`] = `
Array [
  <header
    className="header"
  >
    <h1>
      Workout App
    </h1>
    <div>
      <button
        className="button"
        type="button"
      >
        <a
          className="logIn"
          href="/login"
          onClick={[Function]}
        >
          Log in
        </a>
      </button>
    </div>
  </header>,
  <form
    className="card undefined"
    onSubmit={[Function]}
  >
    <label>
      Forgot Password?
    </label>
    <input
      className="input"
      onChange={[Function]}
      placeholder="Email Address"
      type="email"
      value=""
    />
    <button
      className="button"
      disabled={true}
      type="submit"
    >
      Submit
    </button>
  </form>,
]
`;

exports[`Login page Component routes Sign up link to Register component 1`] = `
Array [
  <header
    className="header"
  >
    <h1>
      Workout App
    </h1>
    <div>
      <button
        className="button"
        type="button"
      >
        <a
          className="logIn"
          href="/login"
          onClick={[Function]}
        >
          Log in
        </a>
      </button>
    </div>
  </header>,
  <form
    className="card undefined"
    onSubmit={[Function]}
  >
    <div>
      <div
        className="noAlert"
      >
        This Email address already exists
      </div>
      <div
        className="noAlert"
      >
        Please enter a valid Email address
      </div>
      <div
        className="noAlert"
      >
        Password should be at least 6 characters. Please try again
      </div>
      <div
        className="noAlert"
      >
        Please enter a valid password
      </div>
    </div>
    <span>
      invalid name
    </span>
    <input
      className="input invalid"
      onChange={[Function]}
      placeholder="First Name"
      type="firstName"
    />
    <input
      className="input"
      onChange={[Function]}
      placeholder="Last Name"
      type="lastName"
    />
    <input
      className="input false"
      onChange={[Function]}
      placeholder="E-Mail"
      type="email"
    />
    <input
      className="input false"
      onChange={[Function]}
      placeholder="Password"
      type="password"
    />
    <button
      className="button"
      disabled={false}
      type="submit"
    >
      Sign Up
    </button>
    <div
      className="signIn"
    >
      Have an account?
      <a
        className="signInLink"
        href="/login"
        onClick={[Function]}
      >
        Sign in
      </a>
    </div>
  </form>,
]
`;

exports[`Login page Login Page renders correctly 1`] = `
Array [
  <header
    className="header"
  >
    <h1>
      Workout App
    </h1>
    <div>
      <button
        className="button"
        type="button"
      >
        <a
          className="logIn"
          href="/login"
          onClick={[Function]}
        >
          Log in
        </a>
      </button>
    </div>
  </header>,
  <form
    className="card undefined"
    onSubmit={[Function]}
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="E-mail.com"
      type="email"
      value=""
    />
    <input
      className="input"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <button
      className="button"
      disabled={false}
      type="submit"
    >
      Log In
    </button>
    <div
      className="signingActivity"
    >
      <a
        className="forgotPass"
        href="/forgot-password"
        onClick={[Function]}
      >
        Forgot Password?
      </a>
      <a
        className="signUp"
        href="/register"
        onClick={[Function]}
      >
        Sign up
      </a>
    </div>
  </form>,
]
`;
